core/hosting/HostWithCoreClrHost/src/SampleHost.cpp

```diff
@@ -0,0 +1,324 @@
+    // This is the list of assemblies that .NET Core can load as
+    // trusted system assemblies (similar to the .NET Framework GAC).
+    // For this host (as with most), assemblies next to CoreCLR will 
+    // be included in the TPA list
+    std::string tpaList;
+    BuildTpaList(runtimePath, ".dll", tpaList);
+
+    // <Snippet3A>
+    // Define CoreCLR properties
+    const char* propertyKeys[] = {
+        "TRUSTED_PLATFORM_ASSEMBLIES",      // Trusted assemblies (like the GAC)
+        "APP_PATHS",                        // Directories to probe for application assemblies
+        // "APP_NI_PATHS",                     // Directories to probe for application native images (not used in this sample)
+        // "NATIVE_DLL_SEARCH_DIRECTORIES",    // Directories to probe for native dlls (not used in this sample)
+    };
+
+    const char* propertyValues[] = {
+        tpaList.c_str(),
+        runtimePath
+    };
+    // </Snippet3>
+
+    //
+    // STEP 4: Start the CoreCLR runtime
+    //
+
+    // <Snippet3>
+    void* hostHandle;
+    unsigned int domainId;
+
+    // This function both starts the .NET Core runtime and creates
+    // the default AppDomain
+    int hr = initializeCoreClr(
+                    runtimePath,        // AppDomain base path
+                    "SampleHost",       // AppDomain friendly name
+                    sizeof(propertyKeys) / sizeof(char*),   // Property count
+                    propertyKeys,       // Property names
+                    propertyValues,     // Property values
+                    &hostHandle,        // Host handle
+                    &domainId);         // AppDomain ID
+    // </Snippet4>
+
+    if (hr >= 0)
+    {
+        printf("CoreCLR started; AppDomain %d created\n", domainId);
+    }
+    else
+    {
+        printf("coreclr_initialize failed - status: 0x%08x\n", hr);
+        return -1;
+    }                  
+
+    //
+    // STEP 5: Create delegate to managed code
+    //
+
+    // <Snippet5>
+    doWork_ptr managedDelegate;    
+    hr = createManagedDelegate(
+            hostHandle, 
+            domainId,
+            "ManagedLibrary, Version=1.0.0.0",
+            "ManagedLibrary.ManagedWorker",
+            "DoWork",
+            (void**)&managedDelegate);
+    // </Snippet5>
+
+    if (hr >= 0)
+    {
+        printf("Managed delegate created\n");
+    }
+    else
+    {
+        printf("coreclr_create_delegate failed - status: 0x%08x\n", hr);
+        return -1;
+    }    
+
+    //
+    // STEP 5.2: Invoking managed delegate
+    //
+
+    // Create sample data for the double[] argument of the managed method to be called
+    double data[4];
+    data[0] = 0; 
+    data[1] = 0.25;
+    data[2] = 0.5;
+    data[3] = 0.75;
+
+    // Invoke the managed delegate and write the returned string to the console
+    const char* ret = managedDelegate("Test job", 5, sizeof(data) / sizeof(double), data, ReportProgressCallback);
```

Comment:
The `ret` string is leaked in this case. For Windows it should be freed with [`CoTaskMemFree()`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-cotaskmemfree) and for non-Windows use [`free()`](https://en.cppreference.com/w/c/memory/free).

Reply:
Good catch! I've fixed this and added a comment pointing out that the host is responsible for freeing strings managed code returns to it.